Test scripts should retry after Docker Hub errors

We often see transient 400 or 504 errors when pushing images to Docker Hub:
```
1af660919603: Pushing
time="2015-06-10T11:59:14-06:00" level="fatal" msg="HTTP code 400 while uploading metadata: {\"error\": \"This image does not belong to the repository\"}" 
```
or
```
f3e0d337c0f9: Pushing
time="2015-06-10T11:20:21-06:00" level="fatal" msg="HTTP code 504 while uploading metadata: unexpected end of JSON input" 
make[1]: *** [push] Error 1
```

Trying again generally succeeds without displaying the error. Ideally the `docker push` parts of our test scripts should catch these errors and simply retry the operation.

--- a/tests/bin/test-integration-ec2.sh
+++ b/tests/bin/test-integration-ec2.sh
@@ -123,9 +123,33 @@ log_phase "etcd available after $WAIT_TIME seconds"
 
 log_phase "Publishing release from source tree"
 
-# TODO: detect where IMAGE_PREFIX=deis/ and DEV_REGISTRY=registry.hub.docker.com
-# and disallow it so we can't pollute the production account.
-make dev-release
+set +e
+trap - ERR
+
+RETRY_COUNT=1
+
+while [ $RETRY_COUNT -le 3 ]; do
+  # TODO: detect where IMAGE_PREFIX=deis/ and DEV_REGISTRY=registry.hub.docker.com
+  # and disallow it so we can't pollute the production account.
+  make dev-release
+  RESULT=$?
+
+  if [ $RESULT -ne 0 ]; then
+    echo "Docker Hub push failed. Attempt $RETRY_COUNT of 3."
+  else
+    break
+  fi
+
+  (( RETRY_COUNT += 1 ))
+done
+
+set -e
+trap dump_logs ERR
+
+if [ $RETRY_COUNT -gt 3 ]; then
+  echo "Docker Hub push failed the maximum number of times, aborting."
+  false
+fi
 
 log_phase "Provisioning Deis"
 

